Apply transform3D updates for model-specific layers

Phipson opened the pull request:
#### bb4229348e91c1a02b0bd8231bcb54a700f51fc3
<pre>
Apply transform3D updates for model-specific layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=296668">https://bugs.webkit.org/show_bug.cgi?id=296668</a>
<a href="https://rdar.apple.com/154223351">rdar://154223351</a>

Reviewed by Ada Chan.

This PR introduces a patch to target the transform3D changes we made specifically
for model layers. Previously, we applied the transform3D to the web view to
scale models such that they would update based on page scale. With this targeted fix,
we ensure that any non-model layers are unaffected by the transform3D changes we apply,
which also means that pages without model elements remain unchanged.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidZoom:]):
(-[WKWebView _willInvalidateDraggedModelWithContainerView:]):
- Routed calls through to the WKContentView to get access to the ModelPresentationManagerProxy
(-[WKWebView _setWebViewTransform3DForModel:]): Deleted.
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
(WebKit::ModelPresentationManagerProxy::setUpModelView):
- Added a call to update the model view layer&apos;s z scale when we first create the layer
(WebKit::ModelPresentationManagerProxy::pageScaleDidChange):
- Added a call to update the model view layer&apos;s z scale when we change the page scale
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setTransform3DForModelView:]):
- Adds a call to the ModelPresentationManagerProxy to update the model view layer based
on the page scale

Canonical link: <a href="https://commits.webkit.org/298100@main">https://commits.webkit.org/298100@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4746eda021f560cca6b9d8f86f4f998fea3e0a0f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33844 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120260 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64856 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34472 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42404 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86711 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27447 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102473 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67095 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26630 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63959 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96815 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123485 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30644 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95545 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95328 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24321 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40464 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18277 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37266 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40998 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46502 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40618 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43918 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/88e566d3665624215b6fcfaa69c2ae6b4a01d0c7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48695

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d20e9365b363c560642485137506a7616954f3d5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48695

etiennesegonzac commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2566,7 +2566,7 @@ - (void)scrollViewDidZoom:(UIScrollView *)scrollView
 
 #if ENABLE(MODEL_PROCESS)
-    [self _setWebViewTransform3DForModel:[scrollView zoomScale]];
+    [_contentView _setWebViewTransform3DForModel:[scrollView zoomScale]];
```
we should rename this since we're not setting the transform on the WebView anymore

etiennesegonzac commented on Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
```diff
@@ -103,6 +104,17 @@
+    for (auto& modelPresentation : m_modelPresentations.values()) {
+        if (RetainPtr modelView = modelPresentation->remoteModelView) {
+            CATransform3D newTransform = [modelView transform3D];
+            newTransform.m33 = newScale;
```
we could add a comment explaining that this is safe because only the pageHostedView is part of the RemoteLayerTree (which can also set the transform)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/80ebaf2212f8eccde5a12142e0845c2c7327fb02. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48695

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4746eda021f560cca6b9d8f86f4f998fea3e0a0f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48695

Phipson committed:
Apply transform3D updates for model-specific layers
https://bugs.webkit.org/show_bug.cgi?id=296668
rdar://154223351

Reviewed by Ada Chan.

This PR introduces a patch to target the transform3D changes we made specifically
for model layers. Previously, we applied the transform3D to the web view to
scale models such that they would update based on page scale. With this targeted fix,
we ensure that any non-model layers are unaffected by the transform3D changes we apply,
which also means that pages without model elements remain unchanged.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidZoom:]):
(-[WKWebView _willInvalidateDraggedModelWithContainerView:]):
- Routed calls through to the WKContentView to get access to the ModelPresentationManagerProxy
(-[WKWebView _setWebViewTransform3DForModel:]): Deleted.
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
(WebKit::ModelPresentationManagerProxy::setUpModelView):
- Added a call to update the model view layer's z scale when we first create the layer
(WebKit::ModelPresentationManagerProxy::pageScaleDidChange):
- Added a call to update the model view layer's z scale when we change the page scale
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setTransform3DForModelView:]):
- Adds a call to the ModelPresentationManagerProxy to update the model view layer based
on the page scale

Canonical link: https://commits.webkit.org/298100@main
